### PR TITLE
WIP RealSphericalHarmonics

### PR DESCRIPTION
**Pull request by TSGut**

### CODE REVIEW

**Comment by TSGut** [comment]
@dlfivefifty, starting this PR to keep track of progress on https://github.com/JuliaApproximation/SphericalHarmonics.jl/issues/8.

File: src/SphericalHarmonics.jl [hunk 1]
```diff
@@ -146,6 +147,18 @@ function getindex(S::SphericalHarmonic{T}, x::ZSphericalCoordinate, K::BlockInde
     exp((lgamma(ℓ+m̃)+lgamma(ℓ-m̃)-2lgamma(ℓ))/2)*sqrt((2ℓ-1)/(4π)) * exp(im*m*x.φ) * sin(θ/2)^m̃ * cos(θ/2)^m̃ * Jacobi{real(T)}(m̃,m̃)[x.z, ℓ-m̃]
 end
 
+function getindex(S::RealSphericalHarmonic{T}, x::ZSphericalCoordinate, K::BlockIndex{1}) where T
+    ℓ = Int(block(K))
+    k = blockindex(K)
+    m = k-ℓ
```
Comment (dlfivefifty) [inline src/SphericalHarmonics.jl:153]: Here is where I would differ from the complex case and instead of `m >0` vs `m < 0` I would do `iseven(k)` vs `isodd(k)`.

@MikaelSlevinsky This differs from your convention. Do you strongly disagree? One argument in favour of Mikael's approach is that it is consistent with FFTW.

That said, I'm planning to make the axes in the complex case indexed from `1-k:k-1` so that we get true negative indexing in which case the index matches `m`. There's an argument for doing this also in the real case. But assigning negative `m` to sine terms just feels wrong to me.

Comment (MikaelSlevinsky) [reply src/SphericalHarmonics.jl:153]: I don't have any strong feelings about any conventions. I chose the conventions that make the transforms carry the least amount of silly flops (even if they're just O(n<sup>2</sup>)), which implies a more maintainable codebase, personally. You can build on this however you feel comfortable!

File: src/SphericalHarmonics.jl [hunk 2]
```diff
@@ -147,16 +147,32 @@ function getindex(S::SphericalHarmonic{T}, x::ZSphericalCoordinate, K::BlockInde
...
+#     m̃ = abs(m)
+#     indepm = (-1)^m̃*exp((lgamma(ℓ-m̃)-lgamma(ℓ+m̃))/2)*sqrt((2ℓ-1)/(2π))*associatedlegendre(m̃)[x.z,ℓ-m̃]
+#     m>0 && return cos(m*x.φ)*indepm
+#     m==0 && return cos(m*x.φ)/sqrt(2)*indepm
+#     m<0 && return sin(m̃*x.φ)*indepm
+# end
+
 function getindex(S::RealSphericalHarmonic{T}, x::ZSphericalCoordinate, K::BlockIndex{1}) where T
+    # starts with m=0, then alternates between sin and cos terms, 
+    # even entries correspond to cos terms, odd entries correspond to sin terms
```
Comment (TSGut) [inline src/SphericalHarmonics.jl:164]: my comment here probably confuses more than it explains, I'll make it make more sense with the next commit. 

It's referring to m rather than the actual index with which we call the element, let's call that k. Even m terms are placed on odd k values by necessity of having the even m=0 in first place.

**Comment by TSGut** [comment]
@dlfivefifty With this, I believe real spherical harmonics now have reached feature parity with your complex spherical harmonics unless I'm missing something. Indexing now also that starts with m=0 and alternates as discussed in https://github.com/JuliaApproximation/SphericalHarmonics.jl/issues/8

I think the provided tests are a bit limited, so I'll probably add some more later.

File: Project.toml [hunk 3]
```diff
@@ -19,12 +19,10 @@ StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
 
 [compat]
 BlockArrays = "0.14"
-ContinuumArrays = "0.4"
```
Comment (dlfivefifty) [inline Project.toml:22]: Why are these deleted?

File: src/SphericalHarmonics.jl [hunk 4]
```diff
@@ -57,6 +56,35 @@ end
 
 getindex(A::SphereTrav, k::Int) = A[findblockindex(axes(A,1), k)]
 
+struct RealSphereTrav{T, AA<:AbstractMatrix{T}} <: AbstractBlockVector{T}
```
Comment (dlfivefifty) [inline src/SphericalHarmonics.jl:59]: Add documentation explaining what this does

File: src/SphericalHarmonics.jl [hunk 5]
```diff
@@ -146,6 +175,33 @@ function getindex(S::SphericalHarmonic{T}, x::ZSphericalCoordinate, K::BlockInde
...
+#     m>0 && return cos(m*x.φ)*indepm
+#     m==0 && return cos(m*x.φ)/sqrt(2)*indepm
+#     m<0 && return sin(m̃*x.φ)*indepm
+# end
+
+function getindex(S::RealSphericalHarmonic{T}, x::ZSphericalCoordinate, K::BlockIndex{1}) where T
+    # starts with m=0, then alternates between sin and cos terms (beginning with sin).
+    ℓ = Int(block(K))
+    m = blockindex(K)-1
+    if m==0
```
Comment (dlfivefifty) [inline src/SphericalHarmonics.jl:194]: This could be `iszero(m)` to make it more consistent

File: src/SphericalHarmonics.jl [hunk 6]
```diff
@@ -146,6 +175,33 @@ function getindex(S::SphericalHarmonic{T}, x::ZSphericalCoordinate, K::BlockInde
...
+        return sqrt((2ℓ-1)/(4*π))*associatedlegendre(0)[x.z,ℓ]
+    elseif isodd(m)
+        m = Int((m+1)/2)
```
Comment (dlfivefifty) [inline src/SphericalHarmonics.jl:197]: Ewww! I think you mean `div(m + 1,2)` or even better `(m+1) ÷ 2`

**Review by dlfivefifty — CHANGES_REQUESTED** [review]
Looks good just a few changes needed

File: Project.toml (hunk 3 above)
Comment (TSGut) [reply Project.toml:22]: I was running into a dependency clash between branches that should now be resolved with your recent tags. I'll re-add the entry in the next commit.

**Comment by dlfivefifty** [comment]
Oh sorry didn't realise merging my PR would close this... can you reopen with a PR into master?

**Comment by TSGut** [comment]
No worries, just a second.
